Add `dotnet new` template to repo

tylerl0706 opened the pull request:
We can ship this to nuget.org so that users can easily create new projects off of this template.



# PowerShellStandard-based PowerShell binary module template

A `dotnet new` template that creates an example PowerShell binary module that uses PowerShellStandard.

```
dotnet new psmodule
```

## Installation

### From nuget.org

> Installation via `dotnet new` using nuget.org coming soon.

### From source

1. Clone the repo.
2. Install the template:

```powershell
dotnet new -i ./src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template
```

## Usage

Once the template is installed, you will see it in your template list:

```
PS> dotnet new -l

Usage: new [options]

Options:
  -h, --help          Displays help for this command.
  -l, --list          Lists templates containing the specified name. If no name is specified, lists all templates.
  -n, --name          The name for the output being created. If no name is specified, the name of the current directory is used.
  -o, --output        Location to place the generated output.
  -i, --install       Installs a source or a template pack.
  -u, --uninstall     Uninstalls a source or a template pack.
  --type              Filters templates based on available types. Predefined values are "project", "item" or "other".
  --force             Forces content to be generated even if it would change existing files.
  -lang, --language   Specifies the language of the template to create.


Templates                                         Short Name       Language          Tags
---------------------------------------------------------------------------------------------------------------------
Console Application                               console          [C#], F#, VB      Common/Console
PowerShell Standard Module                        psmodule         [C#]              Common/Console/PowerShell/Module
Class library                                     classlib         [C#], F#, VB      Common/Library
Unit Test Project                                 mstest           [C#], F#, VB      Test/MSTest
xUnit Test Project                                xunit            [C#], F#, VB      Test/xUnit
ASP.NET Core Empty                                web              [C#], F#          Web/Empty
ASP.NET Core Web App (Model-View-Controller)      mvc              [C#], F#          Web/MVC
ASP.NET Core Web App                              razor            [C#]              Web/MVC/Razor Pages
ASP.NET Core with Angular                         angular          [C#]              Web/MVC/SPA
ASP.NET Core with React.js                        react            [C#]              Web/MVC/SPA
ASP.NET Core with React.js and Redux              reactredux       [C#]              Web/MVC/SPA
ASP.NET Core Web API                              webapi           [C#], F#          Web/WebAPI
global.json file                                  globaljson                         Config
NuGet Config                                      nugetconfig                        Config
Web Config                                        webconfig                          Config
Solution File                                     sln                                Solution
Razor Page                                        page                               Web/ASP.NET
MVC ViewImports                                   viewimports                        Web/ASP.NET
MVC ViewStart                                     viewstart                          Web/ASP.NET
```

To get more details, add the `-h` flag:

```
PS > dotnet new psmodule -h
Usage: new [options]

Options:
  -h, --help          Displays help for this command.
  -l, --list          Lists templates containing the specified name. If no name is specified, lists all templates.
  -n, --name          The name for the output being created. If no name is specified, the name of the current directory is used.
  -o, --output        Location to place the generated output.
  -i, --install       Installs a source or a template pack.
  -u, --uninstall     Uninstalls a source or a template pack.
  --type              Filters templates based on available types. Predefined values are "project", "item" or "other".
  --force             Forces content to be generated even if it would change existing files.
  -lang, --language   Specifies the language of the template to create.


PowerShell Standard Module (C#)
Author: Microsoft Corporation
Options:
  -v|--powershell-standard-version
                                        5.1.0-preview-03    - PowerShell Standard 5.1
                                        3.0.0-preview-02    - PowerShell Standard 3.0
                                    Default: 5.1.0-preview-03

  --no-restore                      If specified, skips the automatic restore of the project on create.
                                    bool - Optional
                                    Default: false
```

To create a template using the defaults:

```
> dotnet new psmodule
The template "PowerShell Standard Module" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on /Users/tylerleonhardt/Downloads/MyProject/MyProject.csproj...
  Restoring packages for /Users/tylerleonhardt/Downloads/MyProject/MyProject.csproj...
  Generating MSBuild file /Users/tylerleonhardt/Downloads/MyProject/obj/MyProject.csproj.nuget.g.props.
  Generating MSBuild file /Users/tylerleonhardt/Downloads/MyProject/obj/MyProject.csproj.nuget.g.targets.
  Restore completed in 275.57 ms for /Users/tylerleonhardt/Downloads/MyProject/MyProject.csproj.

Restore succeeded.
```

Notice that it restores automatically.

You can optionally specify PowerShell Standard V3 by running:

```
dotnet new psmodule --powershell-standard-version 3.0.0-preview-02
```

This template will create:

* `*.csproj` - a project file that uses the same name as the folder it was created in
* `InvokeNoun.cs` - an example PowerShell cmdlet class
* `obj` - a folder generated by the restore

tylerl0706 commented:
Ideally I'd like to work this into the build in some way... but the only way to package this into a `nupkg` is by using `nuget.exe` you ironically can't do it with `dotnet pack`...

tylerl0706 commented:
Also, something that could be expanded on is that we could add a parameter called `--include-tests` that could include a test folder with Pester tests in it.

tylerl0706 commented:
@JamesWTruher: hmm... Now that you mention it I think we can.

```powershell
#install
dotnet new -i "path/to/folder"

dotnet new psmodule
#verify it worked

#cleanup
dotnet new -u "path/to/folder"
```

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template.nuspec:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
+  <metadata>
+    <id>Microsoft.PowerShell.PowerShellStandard.Template</id>
```
I think Module should be in the id as this isn't for PowerShell Standard itself.  Maybe Microsoft.PowerShell.Standard.Module.Template

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template.nuspec:
```diff
@@ -0,0 +1,20 @@
+    <iconUrl>https://secure.gravatar.com/avatar/cfba7016de8e12d5b3017130f75ef8ed</iconUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <description>
+      Creates a PowerShellStandard-based PowerShell binary module.
```
Creates a PowerShell Standard based C# module

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template.nuspec:
```diff
@@ -0,0 +1,20 @@
+    <description>
+      Creates a PowerShellStandard-based PowerShell binary module.
+    </description>
+    <copyright>Copyright Microsoft Corporation 2018. All rights reserved.</copyright>
```
Copyright (c) Microsoft Corporation. All rights reserved.

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/.template.config/template.json:
```diff
@@ -0,0 +1,53 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "Microsoft Corporation",
+  "classifications": [ "Common", "Console", "PowerShell", "Module" ],
```
The module isn't a Console app.  Is classification for the resulting output?  Maybe it should be classlib?

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/.template.config/template.json:
```diff
@@ -0,0 +1,53 @@
+  "name": "PowerShell Standard Module",
+  "identity": "Microsoft.PowerShell.PowerShellStandard.Template",
+  "groupIdentity": "Microsoft.PowerShell.PowerShellStandard.Template",
+  "shortName": "psmodule",
```
Does the casing affect how it's displayed?  If so, use `PSModule`

tylerl0706 replied:
All of the other templates use all lowercase. I don't think we should go against that trend. What do you think?

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/InvokeNoun.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.PowerShell.PowerShellStandard.Template
+{
+    [Cmdlet("Invoke","Noun")]
```
My preference would be to use the `Test` verb instead of `Invoke` as people may tend to leave it as `Invoke` but likely to change `Test`.  Also better to use the enum for the Verb than a string to enforce proper verb usage.

For the Noun, perhaps call it `SampleCmdlet` to make it more obvious they need to change it?

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/InvokeNoun.cs:
```diff
@@ -0,0 +1,50 @@
+namespace Microsoft.PowerShell.PowerShellStandard.Template
+{
+    [Cmdlet("Invoke","Noun")]
+    public class InvokeNoun : PSCmdlet
```
At least with PowerShell, convention is to suffix with Command potentially making it easier to find in complex code.  So `TestSampleCmdletCommand`

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/InvokeNoun.cs:
```diff
@@ -0,0 +1,50 @@
+
+        protected override void BeginProcessing()
+        {
+            WriteObject("Begin!");
```
Maybe use WriteWarning instead of showing how to write strings to the pipeline.  This sample should have some OutputType that does get written to the pipeline to show proper cmdlet design.

tylerl0706 replied:
Can you elaborate on this a bit more? I don't have too much experience with binary modules.

Are you looking for the cmdlet to return some object?

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/InvokeNoun.cs:
```diff
@@ -0,0 +1,50 @@
+                case "Verbose":
+                    WriteVerbose(Value);
+                    break;
+                case "Error":
```
This doesn't match your ValidateSet...

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.PowerShellStandard.Template/Microsoft.PowerShell.PowerShellStandard.Template/InvokeNoun.cs:
```diff
@@ -0,0 +1,50 @@
+            WriteObject("End!");
+        }
+    }
+}
```
add newline

tylerl0706 commented:
Addressed all feedback and added a few tests.

The new cmdlet is:

```
Test-SampleCmdlet
```
with an integer param for `FavoriteNumber` and an option param for `FavoritePet`
```
Test-SampleCmdlet -FavoriteNumber 4
Test-SampleCmdlet -FavoriteNumber 4 -FavoritePet Cat
Test-SampleCmdlet 4 Cat
@(2,3,4) | Test-SampleCmdlet -FavoritePet Cat
```

Tests also verify the cmdlet works.

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.Standard.Module.Template/Microsoft.PowerShell.Standard.Module.Template/TestSampleCmdletCommand.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Microsoft.PowerShell.Standard.Module.Template
+{
+    [Cmdlet("Test","SampleCmdlet")]
```
Replace `"Test"` with `VerbsDiagnostic.Test`

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.Standard.Module.Template/Microsoft.PowerShell.Standard.Module.Template/TestSampleCmdletCommand.cs:
```diff
@@ -0,0 +1,48 @@
+    public class TestSampleCmdletCommand : PSCmdlet
+    {
+        [Parameter(
+            Mandatory=true,
```
add whitespace around the equals sign

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.Standard.Module.Template/Microsoft.PowerShell.Standard.Module.Template/TestSampleCmdletCommand.cs:
```diff
@@ -0,0 +1,48 @@
+
+        protected override void BeginProcessing()
+        {
+            WriteVerbose("Begin!");
```
Add a comment:

```powershell
# this method gets called once for each cmdlet in the pipeline when the pipeline starts executing
```

SteveL-MSFT commented on src/dotnetTemplate/Microsoft.PowerShell.Standard.Module.Template/Microsoft.PowerShell.Standard.Module.Template/TestSampleCmdletCommand.cs:
```diff
@@ -0,0 +1,48 @@
+            });
+        }
+
+        protected override void EndProcessing()
```
Add a comment:

```powershell
# this method will be called once at the end of pipeline execution; if no input is received, this method is not called
